### PR TITLE
Add support for flexible clusters - public/private clusters toggle

### PR DESCRIPTION
**Pull request by Francis-Liu**
This PR adds support for GKE flexible clusters.

If this PR is for Terraform, I acknowledge that I have:

- [ ] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
container: added field `gcp_public_cidrs_access_enabled` and `private_endpoint_subnetwork` to `google_container_cluster`
```

```release-note:enhancement
container: added update support for `enable_private_endpoint` and `enable_private_nodes` in `google_container_cluster`
```

```release-note:enhancement
container: promoted `network_config` in `google_container_node_pool` to GA.
```

```release-note:enhancement
container: added field `enable_private_nodes` in `network_config` to `google_container_node_pool`
```

### CODE REVIEW

**Comment by shuyama1** [comment]
@Francis-Liu Thanks for making the changes. Please let me know if you need help debugging!

**Comment by Francis-Liu** [comment]
@huyama1 Thanks for offering the help! The test errors are expected to be gone after the API rollout tomorrow.

**Comment by Francis-Liu** [comment]
Somebody else has upgraded `google.golang.org/api` to `v0.102.0`. So I have removed my first commit which also does the upgrade.

**Comment by shuyama1** [comment]
Build + Tests failed due to
```
--- FAIL: TestProvider (0.09s)
    provider_test.go:434: err: 1 error occurred:
        	* resource google_container_cluster: peering_name: AtLeastOneOf is for configurable attributes,there's nothing to configure on computed-only field
```
which says we should exclude output-only fields from AtLeastOneOf validation

File: mmv1/third_party/terraform/resources/resource_container_node_pool.go.erb [hunk 1]
```diff
@@ -280,10 +279,15 @@ var schemaNodePool = map[string]*schema.Schema{
 					ForceNew:     true,
 					Description:  `Whether to create a new range for pod IPs in this node pool. Defaults are provided for pod_range and pod_ipv4_cidr_block if they are not specified.`,
 				},
-
+				"enable_private_nodes": {
```
Comment (shuyama1) [inline mmv1/third_party/terraform/resources/resource_container_node_pool.go.erb:282]: Looks like we also add `enable_private_nodes` to `google_container_node_pool` resource. Would you mind adding this to the changelog. Plus, do we aslo need to add this to the doc at mmv1/third_party/terraform/website/docs/r/container_node_pool.html.markdown?

File: mmv1/third_party/terraform/resources/resource_container_node_pool.go.erb [hunk 2]
```diff
@@ -265,7 +265,6 @@ var schemaNodePool = map[string]*schema.Schema{
 		Description: `The Kubernetes version for the nodes in this pool. Note that if this field and auto_upgrade are both specified, they will fight each other for what the node version should be, so setting both is highly discouraged. While a fuzzy version can be specified, it's recommended that you specify explicit versions as Terraform will see spurious diffs when fuzzy versions are used. See the google_container_engine_versions data source's version_prefix field to approximate fuzzy versions in a Terraform-compatible way.`,
 	},
 
-<% unless version == 'ga' -%>
```
[the remark this answers is not in the document]
Comment (Francis-Liu) [reply mmv1/third_party/terraform/resources/resource_container_node_pool.go.erb:268]: Added.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 3]
```diff
@@ -714,6 +714,81 @@ func TestAccContainerCluster_withMasterAuthorizedNetworksConfig(t *testing.T) {
...
+
+	return fmt.Sprintf(`
+resource "google_container_cluster" "with_gcp_public_cidrs_access_enabled" {
+  name               = "%s"
+  location           = "us-central1-a"
+  min_master_version = "1.23"
+  initial_node_count = 1
+
+  master_authorized_networks_config {
+	gcp_public_cidrs_access_enabled = %s
```
[the remark this answers is not in the document]
Comment (Francis-Liu) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:774]: Fixed.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 4]
```diff
@@ -4020,6 +4095,154 @@ resource "google_container_cluster" "regional" {
...
+  name               = "%s"
+  location           = "us-central1-a"
+  min_master_version = "1.23"
+  initial_node_count = 1
+
+  network    = google_compute_network.container_network.name
+  subnetwork = google_compute_subnetwork.container_subnetwork1.name
+
+  private_cluster_config {
+	private_endpoint_subnetwork = google_compute_subnetwork.container_subnetwork2.name
```
[the remark this answers is not in the document]
Comment (Francis-Liu) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4163]: Fixed.

File: mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb [hunk 5]
```diff
@@ -4020,6 +4095,154 @@ resource "google_container_cluster" "regional" {
...
+  location           = "us-central1-a"
+  min_master_version = "1.23"
+  initial_node_count = 1
+
+  master_authorized_networks_config {
+    gcp_public_cidrs_access_enabled = false
+  }
+
+  private_cluster_config {
+	enable_private_endpoint = %s
```
[the remark this answers is not in the document]
Comment (Francis-Liu) [reply mmv1/third_party/terraform/tests/resource_container_cluster_test.go.erb:4224]: Fixed.

File: mmv1/third_party/terraform/resources/resource_container_node_pool.go.erb (hunk 1 above)
Comment (Francis-Liu) [reply mmv1/third_party/terraform/resources/resource_container_node_pool.go.erb:282]: Sure, I have added a note about `enable_private_nodes` to the changelog. For the markdown file, it has been already added.

**Review by shuyama1 — APPROVED** [review]

**Pull request merged**